### PR TITLE
New endpoints

### PR DESCRIPTION
**Pull request by krigar1184**

### CODE REVIEW

File: Test/Test.cs [hunk 1]
```diff
@@ -146,164 +46,480 @@ public void InitClient()
...
             var extras = new Dictionary<string, object>() {{"maxPages", 1}};
-            var response = Toby.CreatePrediction((string)CreateDocResponse["documentId"], Example.ModelName(),
-                                                extras: extras);
-            Console.WriteLine($"CreatePrediction. {response}");
-            var expected = new List<string>(){"documentId", "predictions"};
-            CheckKeys(expected, response);
+            var response = Toby.CreatePrediction(
+                (string)CreateDocResponse["documentId"],
+                Example.ModelId(),
+                extras: extras
```
Comment (magnurud) [inline Test/Test.cs:153]: I suggest we remove extras and add maxpages explicitly

File: Test/Test.cs [hunk 2]
```diff
@@ -146,164 +46,480 @@ public void InitClient()
...
+            var response = Toby.ExecuteTransition(transitionId);
+            CheckKeys(new [] {"transitionId", "executionId", "status"}, response);
+        }
+
+        [TestCase(
+            "running",
+            "las:transition-execution:08b49ae64cd746f384f05880ef5de72f",
+            3,
+            null,
+            "completedBy",
```
Comment (magnurud) [inline Test/Test.cs:300]: sortBy is either startTime or endTime

File: Test/Test.cs [hunk 3]
```diff
@@ -146,164 +46,480 @@ public void InitClient()
...
+            "asc"
```
Comment (magnurud) [inline Test/Test.cs:301]: "ascending" or "descending" are the only valid inputs.

File: Test/Test.cs [hunk 4]
```diff
@@ -146,164 +46,480 @@ public void InitClient()
...
+            );
+            var expectedKeys = new [] {"executions"};
+            CheckKeys(expectedKeys, response);
+        }
+
+        [TestCase(
+            3,
+            null,
+            "completedBy",
+            "asc"
```
Comment (magnurud) [inline Test/Test.cs:329]: Same as above

File: Test/Test.cs [hunk 5]
```diff
@@ -146,164 +46,480 @@ public void InitClient()
...
+            int? maxResults = null,
+            string? nextToken = null,
+            string? sortBy = null,
+            string? order = null
+        ) {
+            var transitionId = $"las:transition:{Guid.NewGuid().ToString()}";
+            var statuses = new List<string>{ "running", "succeeded" };
+            var executionIds = new List<string>{
+                $"las:execution:{Guid.NewGuid().ToString()}",
+                $"las:execution:{Guid.NewGuid().ToString()}"
```
Comment (magnurud) [inline Test/Test.cs:341]: "las:transition-execution:<uuid>" are valud execution-ids, these values should yield an error.

File: Test/Test.cs [hunk 6]
```diff
@@ -146,164 +46,480 @@ public void InitClient()
...
+        };
+
+        [Test, TestCaseSource("UpdateTransitionExecutionSources")]
+        public void TestUpdateTransitionExecution(
+            string status,
+            Dictionary<string, string>? output = null,
+            Dictionary<string, string>? error = null
+        ) {
+            var transitionId = $"las:transition:{Guid.NewGuid().ToString()}";
+            var executionId = $"las:execution:{Guid.NewGuid().ToString()}";
```
Comment (magnurud) [inline Test/Test.cs:368]: same as above

File: Test/Test.cs [hunk 7]
```diff
@@ -146,164 +46,480 @@ public void InitClient()
...
+                "logId",
+                "startTime",
+                "status",
+                "transitionId"
+            }, response);
+        }
+
+        [Test]
+        public void TestCreateUser() {
+            var email = "foo@bar.com";
```
Comment (magnurud) [inline Test/Test.cs:390]: This could be a test-case couldn't it?

File: Test/Test.cs [hunk 8]
```diff
@@ -146,164 +46,480 @@ public void InitClient()
...
+            var response = Toby.CreateUser(email);
+            CheckKeys(new [] {"email", "userId"}, response);
+        }
+
+        [Test]
+        public void TestListUsers() {
+            int maxResults = new Random().Next(1, 100);
```
Comment (magnurud) [inline Test/Test.cs:397]: This variable is not used.

File: Test/Test.cs [hunk 9]
```diff
@@ -146,164 +46,480 @@ public void InitClient()
...
+                "endTime",
+                "transitionExecutions"
+            };
+            CheckKeys(expectedKeys, response);
+        }
+
+        [TestCase(
+            3,
+            null,
+            "completedBy",
```
Comment (magnurud) [inline Test/Test.cs:475]: not a valid value of sortBy

File: Test/Test.cs [hunk 10]
```diff
@@ -146,164 +46,480 @@ public void InitClient()
...
+            "asc"
```
Comment (magnurud) [inline Test/Test.cs:476]: not a valid value of order.

**Review by magnurud — CHANGES_REQUESTED** [review]

File: Test/Test.cs (hunk 3 above)
Comment (krigar1184) [reply Test/Test.cs:301]: Why not allow "asc" and "desc"? It's more typical.

File: Test/Test.cs (hunk 4 above)
Comment (krigar1184) [reply Test/Test.cs:329]: Why not allow "asc" and "desc"? It's more typical.

File: Test/Test.cs (hunk 3 above)
Comment (krigar1184) [reply Test/Test.cs:301]: Either way, it should be made a enum in the oas file.

File: Lucidtech/Las/Client.cs [hunk 11]
```diff
@@ -235,28 +284,386 @@ public object CreateBatch(string description)
...
+                queryParams.Add("maxResults", maxResults.ToString());
+            }
+
+            if (nextToken != null) {
+                queryParams.Add("nextToken", nextToken);
+            }
+
+            RestRequest request = ClientRestRequest(Method.GET, "/transitions", null, queryParams);
+            return ExecuteRequestResilient(RestSharpClient, request);
+        }
```
[the remark this answers is not in the document]
Comment (krigar1184) [reply Lucidtech/Las/Client.cs:398]: It's not duplicate, it's overloaded.

File: Lucidtech/Las/Client.cs [hunk 12]
```diff
@@ -11,6 +11,16 @@
 
 namespace Lucidtech.Las
 {
+    class OptionalParameter {
+        private string? Name { get; set; }
```
Comment (magnurud) [inline Lucidtech/Las/Client.cs:15]: Does it make sense that the name is empty? Doesn't that always need to be a string?

File: Lucidtech/Las/Client.cs [hunk 13]
```diff
@@ -11,6 +11,16 @@
...
+        private string? Value { get; }
+
+        public OptionalParameter(string? name = null, string? value = null) {
```
Comment (magnurud) [inline Lucidtech/Las/Client.cs:18]: same as above

File: Lucidtech/Las/Client.cs [hunk 14]
```diff
@@ -35,11 +45,19 @@ public Client(AmazonCredentials credentials)
...
+        public object CreateAsset(byte[] content, Dictionary<string, string?>? optionalParams) {
             string base64Content = System.Convert.ToBase64String(content);
-            var body = new Dictionary<string, string>(){
+            var body = new Dictionary<string, string?>(){
                 {"content", base64Content}
             };
+
+            if (optionalParams != null) {
+                foreach (KeyValuePair<string, string?> entry in optionalParams) {
+                    Console.WriteLine($"{entry.Key} is {entry.Value} of type {entry.GetType()}");
```
Comment (magnurud) [inline Lucidtech/Las/Client.cs:56]: I assume this line is for debug and will be removed before merge?

File: Lucidtech/Las/Client.cs [hunk 15]
```diff
@@ -194,12 +234,13 @@ public object UpdateDocument(string documentId, List<Dictionary<string, string>>
         }
 
         public object DeleteDocuments(string? consentId = null) {
-            var queryParams = new Dictionary<string, object>() {
+            var queryParams = new Dictionary<string, object?>() {
                 {"consentId", consentId},
             };
```
Comment (magnurud) [inline Lucidtech/Las/Client.cs]: consentId has to be a string, we cannot send in null here.

File: Lucidtech/Las/Client.cs [hunk 16]
```diff
@@ -337,15 +386,24 @@ public object CreateBatch(string description)
...
+        public object UpdateSecret(
+            string secretId,
+            Dictionary<string, string>? data,
+            Dictionary<string, string?>? optionalParams = null
+        ) {
+            var body = new Dictionary<string, object?>();
+
+            if (optionalParams != null) {
+                foreach (KeyValuePair<string, string?> entry in optionalParams) {
+                    Console.WriteLine($"{entry.Key} is {entry.Value} of type {entry.GetType()}");
```
Comment (magnurud) [inline Lucidtech/Las/Client.cs:398]: I don't think we want a print here.

File: Lucidtech/Las/Client.cs [hunk 17]
```diff
@@ -354,16 +412,21 @@ public object CreateBatch(string description)
...
                 {"inputJsonSchema", inputJsonSchema},
                 {"outputJsonSchema", outputJsonSchema},
                 {"transitionType", transitionType},
-                {"name", name},
-                {"description", description}
             };
+
+            if (optionalParams != null) {
+                foreach (KeyValuePair<string, string?> entry in optionalParams) {
+                    Console.WriteLine($"{entry.Key} is {entry.Value} of type {entry.GetType()}");
```
Comment (magnurud) [inline Lucidtech/Las/Client.cs:425]: print

File: Lucidtech/Las/Client.cs [hunk 18]
```diff
@@ -406,23 +469,20 @@ public object CreateBatch(string description)
...
                 {"inputJsonSchema", inputJsonSchema},
-                {"outputJsonSchema", outputJsonSchema},
-                {"name", name},
-                {"description", description}
+                {"outputJsonSchema", outputJsonSchema}
             };
+
+            if (optionalParams != null) {
+                foreach (KeyValuePair<string, string?> entry in optionalParams) {
+                    Console.WriteLine($"{entry.Key} is {entry.Value} of type {entry.GetType()}");
```
Comment (magnurud) [inline Lucidtech/Las/Client.cs:481]: print

File: Lucidtech/Las/Client.cs [hunk 19]
```diff
@@ -441,7 +501,7 @@ public object CreateBatch(string description)
             string? sortBy = null,
             string? order = null
         ) {
-            var queryParams = new Dictionary<string, object> {
+            var queryParams = new Dictionary<string, object?> {
                 {"status", statuses},
                 {"executionId", executionIds},
             };
```
Comment (magnurud) [inline Lucidtech/Las/Client.cs]: status and executionId should not be added unless they are provided (and is a string) by the user

File: Lucidtech/Las/Client.cs [hunk 20]
```diff
@@ -524,21 +584,19 @@ public object CreateBatch(string description)
...
 
-            if (name != null) {
-                body.Add("name", name);
-            }
-
-            if (description != null) {
-                body.Add("description", description);
+            if (optionalParams != null) {
+                foreach (KeyValuePair<string, string?> entry in optionalParams) {
+                    Console.WriteLine($"{entry.Key} is {entry.Value} of type {entry.GetType()}");
```
Comment (magnurud) [inline Lucidtech/Las/Client.cs:597]: print

File: Lucidtech/Las/Client.cs [hunk 21]
```diff
@@ -524,21 +584,19 @@ public object CreateBatch(string description)
...
             Dictionary<string, object> spec,
-            string? name = null,
-            string? description = null,
-            Dictionary<string, string>? errorConfig = null
+            Dictionary<string, string>? errorConfig = null,
+            Dictionary<string, string?>? optionalParams = null
         ) {
             var body = new Dictionary<string, object?>{
                 {"specification", spec},
                 {"errorConfig", errorConfig}
```
Comment (magnurud) [inline Lucidtech/Las/Client.cs:592]: error config should not be provided if it is null.

File: Test/Test.cs [hunk 22]
```diff
@@ -55,10 +55,16 @@ public void Setup()
             CreateDocResponse = JsonSerialPublisher.ObjectToDict<Dictionary<string, object>>(response);
         }
 
-        [Test]
-        public void TestCreateAsset() {
+        [TestCase("name", "description")]
+        [TestCase("", "")]
+        // [TestCase(null, null)] TODO: figure out why this doesn't work
```
Comment (magnurud) [inline Test/Test.cs:60]: For CreateAsset there is a flaw in the API such that it doesn't accept null for name and description. For the others Create and Patch functions that use name and description it should be OK.

File: Test/Test.cs [hunk 23]
```diff
@@ -84,11 +90,15 @@ public void Setup()
             CheckKeys(expectedKeys, response);
         }
 
-        [Test]
-        public void TestUpdateAsset() {
+        [TestCase("name", "description")]
+        [TestCase("", "")]
```
Comment (magnurud) [inline Test/Test.cs:94]: Add a testcase with null

**Review by magnurud — CHANGES_REQUESTED** [review]

File: Lucidtech/Las/Client.cs [hunk 24]
```diff
@@ -235,28 +284,386 @@ public object CreateBatch(string description)
...
+            }
+
+            if (error != null) {
+                body.Add("error", error);
+            }
+            var request = ClientRestRequest(Method.PATCH, url, body);
+            return ExecuteRequestResilient(RestSharpClient, request);
+        }
+
+        public object CreateUser(string email) {
```
[the remark this answers is not in the document]
Comment (krigar1184) [reply Lucidtech/Las/Client.cs:552]: I don't see it in oas file, so I suggest we wait with this.

File: Lucidtech/Las/Client.cs [hunk 25]
```diff
@@ -235,28 +284,386 @@ public object CreateBatch(string description)
...
+            return ExecuteRequestResilient(RestSharpClient, request);
+        }
+
+        public object ListWorkflowExecutions(
+            string workflowId,
+            List<string>? statuses = null,
+            int? maxResults = null,
+            string? nextToken = null,
+            string? sortBy = null,
+            string? order = null
```
[the remark this answers is not in the document]
Comment (krigar1184) [reply Lucidtech/Las/Client.cs:645]: I'll add docstrings later in another PR.

File: Lucidtech/Las/Client.cs (hunk 24 above)
[the remark this answers is not in the document]
Comment (magnurud) [reply Lucidtech/Las/Client.cs:534]: I think it was just added

File: Lucidtech/Las/Client.cs [hunk 26]
```diff
@@ -233,58 +221,14 @@ public object UpdateDocument(string documentId, List<Dictionary<string, string>>
             return ExecuteRequestResilient(RestSharpClient, request);
         }
 
-        public object DeleteDocuments(string? consentId = null) {
+        public object DeleteDocuments(string consentId) {
```
Comment (magnurud) [inline Lucidtech/Las/Client.cs:224]: consentId should be optional.

File: Test/Test.cs (hunk 6 above)
Comment (magnurud) [reply Test/Test.cs:385]: "las:transition-execution:"

**Review by magnurud — APPROVED** [review]

**Pull request merged**